src/main/java/uk/gov/pay/directdebit/mandate/services/OneOffMandateService.java

```diff
@@ -39,23 +40,25 @@ public Transaction create(GatewayAccount gatewayAccount, CreatePaymentRequest cr
 
     @Override
     public void confirm(GatewayAccount gatewayAccount, Mandate mandate, ConfirmMandateRequest confirmMandateRequest) {
-        mandateStateUpdateService.canUpdateStateFor(mandate, DIRECT_DEBIT_DETAILS_CONFIRMED);
-
-        Transaction transaction = transactionService
-                .findTransactionForExternalId(confirmMandateRequest.getTransactionExternalId());
-
-        OneOffConfirmationDetails oneOffConfirmationDetails = paymentProviderFactory
-                .getCommandServiceFor(gatewayAccount.getPaymentProvider())
-                .confirmOneOffMandate(mandate,
-                        new BankAccountDetails(
-                                confirmMandateRequest.getAccountNumber(),
-                                confirmMandateRequest.getSortCode()),
-                        transaction
-                );
-
-        mandateStateUpdateService.confirmedOneOffDirectDebitDetailsFor(oneOffConfirmationDetails.getMandate());
-        transactionService.oneOffPaymentSubmittedToProviderFor(
-                transaction,
-                oneOffConfirmationDetails.getChargeDate());
+        if (mandateStateUpdateService.canUpdateStateFor(mandate, DIRECT_DEBIT_DETAILS_CONFIRMED)) {
```

Comment:
I wonder if we could do away with `canUpdateStateFor(…)` altogether and have something like:

```java
mandateStateUpdateService.getNextStateForEvent(mandate.getState(), DIRECT_DEBIT_DETAILS_CONFIRMED).ifPresentOrElse(nextState -> {
    //…
}, () -> throw new InvalidStateTransitionException(…));
```

Reply:
Like the idea but it doesn't appear possible to `throw` from a `Runnable` which is the second argument to `ifPresentOrElse`. Really need an `ifPresentOrElseThrow`, both exist individually but not together (`ifPresent` itself just returns `void`).